File: Algorithms/src/hard/DivisorDilemma.swift
import Foundation

// Fast IO
final class FastScanner {
    private var data:[UInt8] = Array(FileHandle.standardInput.readDataToEndOfFile()) + [0]
    private var idx: Int = 0
    func readInt() -> Int {
        var num = 0
        var sign = 1
        while data[idx] == 10 || data[idx] == 13 || data[idx] == 32 { idx += 1 }
        if data[idx] == 45 { sign = -1; idx += 1 }
        while data[idx] >= 48 {
            num = num * 10 + Int(data[idx] - 48)
            idx += 1
        }
        return num * sign
    }
}

// Iterative Segment Tree
final class SegmentTree {
    let n: Int
    var t: [Int64]

    init(_ n: Int) {
        self.n = n
        self.t = Array(repeating: 0, count: 2 * n)
    }

    func update(_ pos: Int, _ val: Int64) {
        var p = pos + n
        t[p] = val
        while p > 1 {
            p >>= 1
            t[p] = t[p << 1] + t[p << 1 | 1]
        }
    }

    func query(_ l: Int, _ r: Int) -> Int64 {
        var l = l + n
        var r = r + n
        var res: Int64 = 0
        while l < r {
            if (l & 1) == 1 { res += t[l]; l += 1 }
            if (r & 1) == 1 { r -= 1; res += t[r] }
            l >>= 1
            r >>= 1
        }
        return res
    }
}

// Main
let scanner = FastScanner()
let MAXN = 2_000_001

// Sum of divisors
var sumOfDiv = Array(repeating: Int64(0), count: MAXN)
for i in 1..<MAXN {
    var j = i
    while j < MAXN {
        sumOfDiv[j] += Int64(i)
        j += i
    }
}

// Order by sum of divisors
var orderArr = [(Int64, Int)]()
orderArr.reserveCapacity(MAXN)
for i in 1..<MAXN {
    orderArr.append((sumOfDiv[i], i))
}
orderArr.sort { $0.0 < $1.0 }

var order = Array(repeating: 0, count: MAXN)
for i in 0..<orderArr.count {
    order[orderArr[i].1] = i
}

// Read queries
let Q = scanner.readInt()
var queries = [(Int, Int, Int)]()   // (N, M, index)
queries.reserveCapacity(Q)

for i in 0..<Q {
    let N = scanner.readInt()
    let M = scanner.readInt()
    queries.append((N, M, i))
}

// Sort descending by N
queries.sort { $0.0 > $1.0 }

let segCount = SegmentTree(MAXN)
let segSum   = SegmentTree(MAXN)

var ans = Array(repeating: Int64(0), count: Q)

var qi = queries.count - 1

for i in 1..<MAXN {
    let pos = order[i]
    segCount.update(pos, 1)
    segSum.update(pos, sumOfDiv[i])

    while qi >= 0 && queries[qi].0 == i {
        let (_, M, idx) = queries[qi]
        qi -= 1

        var low = 0
        var high = MAXN - 1
        var best = -1

        while low <= high {
            let mid = (low + high) >> 1
            if segCount.query(mid, MAXN) >= Int64(M) {
                best = mid
                low = mid + 1
            } else {
                high = mid - 1
            }
        }

        ans[idx] = segSum.query(best, MAXN)
    }
}

// Output
var output = ""
for v in ans {
    output += "\(v)\n"
}
print(output, terminator: "")

File: Algorithms/src/hard/DivisorDilemma.rs
use std::io::{self, Read};

struct SegTree {
    n: usize,
    t: Vec<i64>,
}

impl SegTree {
    fn new(n: usize) -> Self {
        SegTree {
            n,
            t: vec![0; 2 * n],
        }
    }

    fn update(&mut self, mut p: usize, v: i64) {
        p += self.n;
        self.t[p] = v;
        while p > 1 {
            p >>= 1;
            self.t[p] = self.t[p << 1] + self.t[p << 1 | 1];
        }
    }

    fn query(&self, mut l: usize, mut r: usize) -> i64 {
        let mut res = 0;
        l += self.n;
        r += self.n;
        while l < r {
            if (l & 1) == 1 {
                res += self.t[l];
                l += 1;
            }
            if (r & 1) == 1 {
                r -= 1;
                res += self.t[r];
            }
            l >>= 1;
            r >>= 1;
        }
        res
    }
}

fn main() {
    // ---------- Fast Input ----------
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut it = input.split_whitespace();

    const MAXN: usize = 2_000_001;

    // ---------- Sum of Divisors ----------
    let mut sum_of_div = vec![0i64; MAXN];
    for i in 1..MAXN {
        let mut j = i;
        while j < MAXN {
            sum_of_div[j] += i as i64;
            j += i;
        }
    }

    // ---------- Order by sum of divisors ----------
    let mut pairs: Vec<(i64, usize)> = (1..MAXN)
        .map(|i| (sum_of_div[i], i))
        .collect();

    pairs.sort_unstable();

    let mut order = vec![0usize; MAXN];
    for (idx, &(_, v)) in pairs.iter().enumerate() {
        order[v] = idx;
    }

    // ---------- Read Queries ----------
    let q: usize = it.next().unwrap().parse().unwrap();
    let mut queries = Vec::with_capacity(q);
    for i in 0..q {
        let n: usize = it.next().unwrap().parse().unwrap();
        let m: i64 = it.next().unwrap().parse().unwrap();
        queries.push((n, m, i));
    }

    // Sort descending by N
    queries.sort_unstable_by(|a, b| b.0.cmp(&a.0));

    let mut ans = vec![0i64; q];

    // ---------- Segment Trees ----------
    let mut seg_count = SegTree::new(MAXN);
    let mut seg_sum = SegTree::new(MAXN);

    let mut qi: isize = queries.len() as isize - 1;

    // ---------- Process ----------
    for i in 1..MAXN {
        let pos = order[i];
        seg_count.update(pos, 1);
        seg_sum.update(pos, sum_of_div[i]);

        while qi >= 0 && queries[qi as usize].0 == i {
            let (_, m, idx) = queries[qi as usize];
            qi -= 1;

            let mut low = 0usize;
            let mut high = MAXN - 1;
            let mut best = 0usize;

            while low <= high {
                let mid = (low + high) >> 1;
                if seg_count.query(mid, MAXN) >= m {
                    best = mid;
                    low = mid + 1;
                } else {
                    if mid == 0 { break; }
                    high = mid - 1;
                }
            }

            ans[idx] = seg_sum.query(best, MAXN);
        }
    }

    // ---------- Output ----------
    let mut out = String::new();
    for v in ans {
        out.push_str(&format!("{}\n", v));
    }
    print!("{}", out);
}
